### PR TITLE
test(quality): forbid empty tests and fake assertions

### PR DESCRIPTION
**Pull request by dimileeh**
Automatically opened by AWF workspace `ws_1aacbc2dd884497d8eab8e7d` (agent: `codex`, model: `gpt-5.5`, effort: `xhigh`).

### Task
Implement the P0 test-quality guardrail slice: forbid empty tests, fake assertions, and broad monkeypatching that skips behavior under test.

Acceptance:
- Add a lightweight static test/quality checker that runs in tests and flags obviously empty tests, assert True/False placeholders, pytest.skip-only tests without a real condition, and broad monkeypatches of the entire behavior under test.
- Keep the rule focused to avoid noisy false positives; document allowed escape hatches with explicit rationale.
- Add tests for the checker itself using small fixture files.
- Wire it into AWF self-test validation only if it is stable and low-noise.

Strict TDD requirements:
1. Write failing tests first where the codebase supports it.
2. Implement the smallest production change needed.
3. Preserve existing public API response compatibility unless the task explicitly adds fields.
4. Do not lower coverage thresholds, workspace coverage requirements, or quality gates.
5. Do not manually merge PRs. AWF owns push, PR creation, monitoring, and auto-merge.

Validation:
uv run --python 3.12 --extra dev pytest tests/unit -q
uv run --python 3.12 --extra dev ruff check src/awf tests
uv run --python 3.12 --extra dev mypy src/awf

---
Validation: 7 profile command(s) passed inside the workspace container.

### CODE REVIEW

File: src/awf/control/test_quality_guardrails.py [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+            continue
+        if path.suffix == ".py" and not _is_excluded(path, exclude_globs):
+            yield path
+
+
+def _is_excluded(path: Path, exclude_globs: Sequence[str]) -> bool:
+    names = {path.as_posix()}
+    with suppress(ValueError):
+        names.add(path.resolve().relative_to(Path.cwd().resolve()).as_posix())
+    return any(fnmatch.fnmatch(name, pattern) for name in names for pattern in exclude_globs)
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/test_quality_guardrails.py]: fixed in commit 23da88de4ea8b94cfdd4bb6c4054624357652cf4

File: src/awf/control/test_quality_guardrails.py [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+            )
+        )
+    return hatches
+
+
+def _invalid_escape_reason(code: str, reason: str) -> str | None:
+    if code not in _SUPPRESSIBLE_CODES:
+        return f"test-quality escape hatch uses unknown rule code {code!r}"
+    normalized = " ".join(reason.lower().split())
+    if len(reason) < 24:
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/test_quality_guardrails.py]: fixed in commit ffb55386113d93a417920fcb40a7f9d0a0dc0c5d

File: src/awf/control/test_quality_guardrails.py [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+def _is_excluded(path: Path, exclude_globs: Sequence[str]) -> bool:
+    names = {path.as_posix()}
+    with suppress(ValueError):
+        names.add(path.resolve().relative_to(Path.cwd().resolve()).as_posix())
+    return any(fnmatch.fnmatch(name, pattern) for name in names for pattern in exclude_globs)
+
+
+def _scan_file(path: Path) -> list[TestQualityViolation]:
+    source = path.read_text(encoding="utf-8")
+    tree = ast.parse(source, filename=str(path))
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/test_quality_guardrails.py]: fixed in commit 40de86fcd68c586cf5178b5cad42d1768ff758e6

File: tests/unit/control/test_test_quality_guardrails_self.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+REPO_ROOT = Path(__file__).resolve().parents[3]
+
+
+@pytest.mark.unit
+def test_awf_test_suite_has_no_test_quality_guardrail_violations() -> None:
+    violations = scan_test_quality(
+        [REPO_ROOT / "tests" / "unit", REPO_ROOT / "tests" / "integration"],
+        exclude_globs=("tests/fixtures/**",),
+    )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply tests/unit/control/test_test_quality_guardrails_self.py:18]: fixed in commit e76787f37301cb08b85e409386362b37bbf5174f

File: src/awf/control/test_quality_guardrails.py [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+    for hatch in hatches:
+        if cast(str, violation.code) != hatch.code:
+            continue
+        if hatch.line in {violation.line, violation.line - 1}:
+            return True
+        if violation.test_start <= hatch.line <= violation.test_end:
+            return True
+        if hatch.line == violation.test_start - 1:
+            return True
+    return False
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/test_quality_guardrails.py:647]: fixed in commit 23e0bddd1cae12b632213f6e51229ad86b4f0bac
